theano/gpuarray/type.py

```diff
@@ -218,9 +223,13 @@ def filter(self, data, strict=False, allow_downcast=None):
               (allow_downcast is None and
                type(data) == float and
                self.dtype == config.floatX)):
-            data = gpuarray.array(data, dtype=self.typecode, copy=False,
-                                  ndmin=len(self.broadcastable),
-                                  context=self.context)
+            if not isinstance(data, gpuarray.GpuArray):
+                data = numpy.array(data, dtype=self.dtype, copy=False,
+                                   ndmin=len(self.broadcastable))
+            else:
+                data = gpuarray.array(data, dtype=self.typecode, copy=False,
+                                      ndmin=len(self.broadcastable),
+                                      context=self.context)
```

Comment:
Just something like:

shared = theano.shared(data, target='cuda')
old_container = shared.get_value(borrow=True, return_internal_type=True)
new_data = np...
shared.set_vaue(new_data)
assert old_container is shared.get_value(borrow=True, return_internal_type=True)